### ISSUE TITLE
Is it possible to have two nodes created in one Notebook?

### ISSUE THREAD

**Issue by pitosalas** [RoboStack/jupyter-ros#27]
I am still trying to figure out the threading model of this approach.

**Comment by wolfv** [comment]
Hi, I think it would be more common to have one node per notebook (maybe the node should actually be started automatically with jupyter-ros...)

But you can definitely have more than one subscriber (runs async in it's own thread) or publisher in the same notebook. For publisher, I had a small decorator to run the publisher in a thread so that you can do something like:

```
@rate(10)
rospy.publish("/mytopic", my_msg)
```

but this part of the code isn't ready yet. If you want to work on something like this that would be awesome.

What's your use case for two nodes?

**Comment by pitosalas** [comment]
Hi

The use case is teaching: have a simple node that is publishing and another simple nose that subscribes and print. The two are used together in explaining the basics.

It’s probably the most common first example. Basically it’s these two, in one notebook.





You say that each node has its own thread, but I don’t understand. How does notebook or jupyter-ros know where one node ends and the new one starts? 

Also I don’t understand decorators or the example you cite below.

Pito Salas
Brandeis Computer Science
Volen 134

**Comment by pitosalas** [comment]
Any response to these questions. A clarificaiton of the threading and node creation model would be very useful.

**Comment by wolfv** [comment]
yeah so this would be most simple to do in either two seperate notebooks (very convenient with JupyterLab since you can just create two tabs and put them next to each other) or you can open two Notebooks in two browser tabs.

Also you don't *need* two nodes for a subscriber and publisher.
You can just create the subscriber in the notebook first, and then later create the publisher. 
The rospy subscriber *automatically* runs in a (Python) thread.

However, the publisher doesn't. You need to manually create a thread, or use the following magic (that has not yet been released):

https://github.com/RoboStack/jupyter-ros/blob/master/jupyros/ipy.py

That would allow you to create a cell like this:

```
%thread_cell

for i in range(0, 1000):
    print("Hello from a thread")
    rospy.publish("...")
```

But it hasn't been released because, IIRC, there were some problems. If you have some PR to make this functionality better that would be awesome! :)
w

**Comment by pitosalas** [comment]
Hi Wolf,

> Also you don't need two nodes for a subscriber and publisher.
> 

I know but I am trying to teach the concept and that is the most natural way to explain it as a first example. But if its not possible then that’s fine/
> If you have some PR to make this functionality better that would be awesome! :
> 
I would be happy to but I don’t understand how it all works yet :)

Pito

**Comment by wolfv** [comment]
As far as I know you also can't have two nodes in a single python script ... So it's not even a limitation of Jupyter, but rather of rospy that you need two python processes for two real nodes.

**Comment by pitosalas** [comment]
Yes but one of the advantages of using Jupiter Notebook would be to allow better exposition of the concepts in one notebook. Anyway, it’s fine.

ps. Would you prefer I correspond to you directly or continue to post issues that are really questions? Also I just checked out the Jupiter Support in the vscode python package from Microsoft. It’s a very nice implementation. I will be trying jupyter-ros with it and see how it goes.

Pito Salas
Brandeis Computer Science
Volen 134
